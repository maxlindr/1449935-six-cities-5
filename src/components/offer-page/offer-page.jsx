import React from 'react';
import PropTypes from 'prop-types';
import {offerPropTypes, reviewPropTypes} from '../../prop-types/prop-types';
import PageHeader from '../page-header/page-header';
import RatingStars, {RatingStarsType} from '../rating-stars/rating-stars';
import BookmarkToggle, {BookmarkToggleType} from '../bookmark-toggle/bookmark-toggle';
import OfferReviewsSection from '../offer-reviews-section/offer-reviews-section';
import OfferEssentials from '../offer-essentials/offer-essentials';
import OfferFeatures from '../offer-features/offer-features';
import OfferGallery from '../offer-gallery/offer-gallery';
import NearPlaces from '../near-places/near-places';
import OfferHost from '../offer-host/offer-host';
import withLeafletMap from '../../hocs/with-leaflet-map/with-leaflet-map';
import cityMapFactory, {CityMapType} from '../city-map-factory/city-map-factory';
import withExtraOfferData from '../../hocs/with-extra-offer-data/with-extra-offer-data';
import withAlertDialog from '../../hocs/with-alert-dialog/with-alert-dialog';

const MAX_VISIBLE_OFFERS_ON_MAP = 4;

const CityMap = withLeafletMap(cityMapFactory(CityMapType.OFFER));

const PREMIUM_MARK_ELEMENT = (
  <div className="property__mark">
    <span>Premium</span>
  </div>
);

const OfferPage = (props) => {
  const {offer, reviews, offers} = props;
  const {photos, premium, title, favorite, rating, price, features, host, description} = offer;

  const offerReviews = reviews
    ? reviews.slice().sort((a, b) => b.date.getTime() - a.date.getTime())
    : null;

  return (
    <div className="page">
      <PageHeader />

      <main className="page__main page__main--property">
        <section className="property">
          <OfferGallery photos={photos} />

          <div className="property__container container">
            <div className="property__wrapper">
              {premium ? PREMIUM_MARK_ELEMENT : null}

              <div className="property__name-wrapper">
                <h1 className="property__name">{title}</h1>
                <BookmarkToggle
                  offer={offer}
                  type={BookmarkToggleType.OFFER}
                  active={favorite}
                />
              </div>

              <RatingStars type={RatingStarsType.OFFER} rating={rating} />
              <OfferEssentials offer={offer} />

              <div className="property__price">
                <b className="property__price-value">â‚¬{price}</b>
                <span className="property__price-text">&nbsp;night</span>
              </div>

              <OfferFeatures features={features} />
              <OfferHost host={host} description={description} />
              <OfferReviewsSection offerId={offer.id} reviews={offerReviews} />
            </div>
          </div>

          <CityMap
            activeOffer={offer}
            offers={[offer, ...offers]}
            maxOffers={MAX_VISIBLE_OFFERS_ON_MAP}
          />
        </section>

        <div className="container">
          {offers.length > 0 ? <NearPlaces offers={offers} /> : null}
        </div>
      </main>
    </div>
  );
};

OfferPage.propTypes = {
  offer: offerPropTypes.isRequired,
  offers: PropTypes.arrayOf(offerPropTypes).isRequired,
  reviews: PropTypes.arrayOf(reviewPropTypes),
};

export {OfferPage};
export default withExtraOfferData(withAlertDialog(OfferPage));
