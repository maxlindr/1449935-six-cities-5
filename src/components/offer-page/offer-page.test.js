import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import {OfferPage} from './offer-page';
import offerMock from '../../../__mocks__/offer-mock';

const renderer = new ShallowRenderer();

test(`OfferPage должен корректно отображаться`, () => {
  renderer.render(
      <OfferPage
        offer={offerMock}
        activeOffer={null}
        offers={[]}
        reviews={null}
        onUpdate={() => {}}
      />
  );

  const result = renderer.getRenderOutput();
  expect(result).toMatchSnapshot();
});
