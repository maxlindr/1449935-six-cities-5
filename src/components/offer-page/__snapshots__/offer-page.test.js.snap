// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferPage должен корректно отображаться 1`] = `
<div
  className="page"
>
  <Memo(Connect(PageHeader)) />
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <Memo(OfferGallery)
        photos={
          Array [
            "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg",
            "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg",
            "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/9.jpg",
            "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg",
            "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/4.jpg",
            "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/19.jpg",
            "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg",
            "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg",
            "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/2.jpg",
            "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/11.jpg",
            "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/1.jpg",
            "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/17.jpg",
            "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/5.jpg",
            "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg",
          ]
        }
      />
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              The house among olive 
            </h1>
            <Memo(Connect(WithUpdateOfferOnFavoriteToggle))
              active={false}
              offer={
                Object {
                  "bedrooms": 3,
                  "description": "Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs.",
                  "favorite": false,
                  "features": Array [
                    "Laptop friendly workspace",
                    "Washer",
                    "Towels",
                    "Fridge",
                    "Air conditioning",
                    "Baby seat",
                    "Breakfast",
                  ],
                  "host": Object {
                    "avatar": "img/avatar-angelina.jpg",
                    "id": 25,
                    "name": "Angelina",
                    "super": true,
                  },
                  "id": "1",
                  "location": Object {
                    "city": Object {
                      "coordinates": Array [
                        50.938361,
                        6.959974,
                      ],
                      "name": "Cologne",
                      "zoom": 13,
                    },
                    "coordinates": Array [
                      50.957361,
                      6.9509739999999995,
                    ],
                    "zoom": 16,
                  },
                  "maxAdults": 4,
                  "photos": Array [
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/9.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/4.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/19.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/2.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/11.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/1.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/17.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/5.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg",
                  ],
                  "premium": false,
                  "price": 493,
                  "rating": 3.8,
                  "thumbnail": "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg",
                  "title": "The house among olive ",
                  "type": "house",
                }
              }
              type="property"
            />
          </div>
          <Memo(RatingStars)
            rating={3.8}
            type="property"
          />
          <Memo(OfferEssentials)
            offer={
              Object {
                "bedrooms": 3,
                "description": "Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs.",
                "favorite": false,
                "features": Array [
                  "Laptop friendly workspace",
                  "Washer",
                  "Towels",
                  "Fridge",
                  "Air conditioning",
                  "Baby seat",
                  "Breakfast",
                ],
                "host": Object {
                  "avatar": "img/avatar-angelina.jpg",
                  "id": 25,
                  "name": "Angelina",
                  "super": true,
                },
                "id": "1",
                "location": Object {
                  "city": Object {
                    "coordinates": Array [
                      50.938361,
                      6.959974,
                    ],
                    "name": "Cologne",
                    "zoom": 13,
                  },
                  "coordinates": Array [
                    50.957361,
                    6.9509739999999995,
                  ],
                  "zoom": 16,
                },
                "maxAdults": 4,
                "photos": Array [
                  "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg",
                  "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg",
                  "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/9.jpg",
                  "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg",
                  "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/4.jpg",
                  "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/19.jpg",
                  "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg",
                  "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg",
                  "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/2.jpg",
                  "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/11.jpg",
                  "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/1.jpg",
                  "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/17.jpg",
                  "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/5.jpg",
                  "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg",
                ],
                "premium": false,
                "price": 493,
                "rating": 3.8,
                "thumbnail": "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg",
                "title": "The house among olive ",
                "type": "house",
              }
            }
          />
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              493
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <Memo(OfferFeatures)
            features={
              Array [
                "Laptop friendly workspace",
                "Washer",
                "Towels",
                "Fridge",
                "Air conditioning",
                "Baby seat",
                "Breakfast",
              ]
            }
          />
          <Memo(OfferHost)
            description="Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs."
            host={
              Object {
                "avatar": "img/avatar-angelina.jpg",
                "id": 25,
                "name": "Angelina",
                "super": true,
              }
            }
          />
          <Memo(Connect(Component))
            offerId="1"
            reviews={null}
          />
        </div>
      </div>
      <Memo(Connect(WithLeafletMap))
        activeOffer={
          Object {
            "bedrooms": 3,
            "description": "Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs.",
            "favorite": false,
            "features": Array [
              "Laptop friendly workspace",
              "Washer",
              "Towels",
              "Fridge",
              "Air conditioning",
              "Baby seat",
              "Breakfast",
            ],
            "host": Object {
              "avatar": "img/avatar-angelina.jpg",
              "id": 25,
              "name": "Angelina",
              "super": true,
            },
            "id": "1",
            "location": Object {
              "city": Object {
                "coordinates": Array [
                  50.938361,
                  6.959974,
                ],
                "name": "Cologne",
                "zoom": 13,
              },
              "coordinates": Array [
                50.957361,
                6.9509739999999995,
              ],
              "zoom": 16,
            },
            "maxAdults": 4,
            "photos": Array [
              "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg",
              "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg",
              "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/9.jpg",
              "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg",
              "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/4.jpg",
              "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/19.jpg",
              "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg",
              "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg",
              "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/2.jpg",
              "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/11.jpg",
              "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/1.jpg",
              "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/17.jpg",
              "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/5.jpg",
              "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg",
            ],
            "premium": false,
            "price": 493,
            "rating": 3.8,
            "thumbnail": "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg",
            "title": "The house among olive ",
            "type": "house",
          }
        }
        maxOffers={4}
        offers={
          Array [
            Object {
              "bedrooms": 3,
              "description": "Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs.",
              "favorite": false,
              "features": Array [
                "Laptop friendly workspace",
                "Washer",
                "Towels",
                "Fridge",
                "Air conditioning",
                "Baby seat",
                "Breakfast",
              ],
              "host": Object {
                "avatar": "img/avatar-angelina.jpg",
                "id": 25,
                "name": "Angelina",
                "super": true,
              },
              "id": "1",
              "location": Object {
                "city": Object {
                  "coordinates": Array [
                    50.938361,
                    6.959974,
                  ],
                  "name": "Cologne",
                  "zoom": 13,
                },
                "coordinates": Array [
                  50.957361,
                  6.9509739999999995,
                ],
                "zoom": 16,
              },
              "maxAdults": 4,
              "photos": Array [
                "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg",
                "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg",
                "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/9.jpg",
                "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg",
                "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/4.jpg",
                "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/19.jpg",
                "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg",
                "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg",
                "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/2.jpg",
                "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/11.jpg",
                "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/1.jpg",
                "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/17.jpg",
                "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/5.jpg",
                "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg",
              ],
              "premium": false,
              "price": 493,
              "rating": 3.8,
              "thumbnail": "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg",
              "title": "The house among olive ",
              "type": "house",
            },
          ]
        }
      />
    </section>
    <div
      className="container"
    />
  </main>
</div>
`;
